### PR TITLE
LS-5226: Add support for Proto over HTTP

### PR DESCRIPTION
**Pull request by frenchfrywpepper**
- Fixed an issue in examples where argparser will not parse bool
as expected (use_tls)
- Added options to Tracer use_thrift or use_http
- Created Converter abstract class, either a Thrift or an HTTP
converter will be used
- Implemented HTTPConnection

### CODE REVIEW

File: lightstep/converter.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+from abc import ABCMeta, abstractmethod
+
+
+class Converter(object):
+    """Converter is a simple abstract interface for converting span data to wire compatible formats for the Satellites.
+    """
+
+    __metaclass__ = ABCMeta
```
Comment (carlosalberto) [inline lightstep/converter.py]: IIRC this won't work against Python 3.x - maybe use `six` to handle the metaclass case?

Comment (frenchfrywpepper) [reply lightstep/converter.py]: hmm, the tests are running 2.7 and 3.4, is that a false positive?

Comment (frenchfrywpepper) [reply lightstep/converter.py]: I also ran the trivial example with 3.4, worked without any issues.

File: lightstep/http_connection.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
+""" Connection class establishes HTTP connection with server.
+    Utilized to send Proto Report Requests.
+"""
+import threading
+import requests
+
+from lightstep.collector_pb2 import ReportResponse
+
+CONSECUTIVE_ERRORS_BEFORE_RECONNECT = 200
```
[the remark this answers is not in the document]
Comment (jmacd) [reply lightstep/http_connection.py]: I see it was probably copied from the Thrift code path. I suspect this behavior is not as close to our more modern clients than we would like it to be. It's probably good not to interfere with this PR, but maybe worth noting that this is not ideal.  In the Go HTTP client for example, we make one request without setting an explicit timeout, meaning (I think) we use the `http.Client` timeout by default. If the request fails, we try to re-integrate those spans into the buffer (possibly dropping some) but do not repeat the request.

It looks like here, if we're going to repeat the request 200 times, the data may be very stale by the time it actually is received.

[the remark this answers is not in the document]
Comment (jmacd) [reply lightstep/http_connection.py]: note: I see there are no timeout options anywhere in the python tracer. I'm sure Thrift didn't really expose that as an option, but it should be something we can do for an HTTP request.  In the Go tracer library, this is named `Options.ReportTimeout`.

[the remark this answers is not in the document]
Comment (frenchfrywpepper) [reply lightstep/http_connection.py]: I removed the retry stuff and added a timeout.

File: lightstep/http_connection.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    def open(self):
+        """Establish HTTP connection to the server.
+        """
+        pass
+
+    # May throw an Exception on failure.
+    def report(self, *args, **kwargs):
+        """Report to the server."""
+        # Notice the annoying case change on the method name. I chose to stay
```
[the remark this answers is not in the document]
Comment (frenchfrywpepper) [reply lightstep/http_connection.py]: Deleted.

**Comment by aliceadelef** [comment]
LGTM but I am by no means a python expert so may want other 👁 s

File: examples/http/context_in_headers.py [hunk 4]
```diff
@@ -121,18 +122,23 @@ def lightstep_tracer_from_args():
...
+    parser.add_argument('--no_tls', help='Disable TLS for reporting',
+                        dest="no_tls", action='store_true')
     parser.add_argument('--component_name', help='The LightStep component name',
             default='TrivialExample')
     args = parser.parse_args()
 
+    if args.no_tls:
+        collector_encryption = 'none'
+    else:
+        collector_encryption = 'tls'
```
[the remark this answers is not in the document]
Comment (frenchfrywpepper) [reply examples/http/context_in_headers.py]: The parameter is true/false so I have to do some kind of conversion from false to none.

[the remark this answers is not in the document]
Comment (aliceadelef) [reply examples/http/context_in_headers.py]: ohhhh gotcha

File: setup.py [hunk 5]
```diff
@@ -11,7 +11,8 @@
                       'jsonpickle',
                       'six',
                       'basictracer>=2.2,<2.3',
-                      'googleapis-common-protos==1.5.3'],
+                      'googleapis-common-protos==1.5.3',
+                      'requests==2.19.1'],
```
Comment (joeblubaugh) [inline setup.py]: Does this have to be a hard equality? Can we be more flexible with the requirement here, given that requests could be used in the enclosing application too?

**Pull request merged**